zipkin-autoconfigure/storage-cassandra3/src/main/java/zipkin/autoconfigure/storage/cassandra3/brave/TracedSession.java

```diff
@@ -11,204 +11,182 @@
+
+    static TracedSession create(final Session session, Brave brave, SpanCollector collector) {
+        return new TracedSession(session, brave, collector);
     }
-    try {
-      return method.invoke(target, args);
-    } catch (InvocationTargetException e) {
-      if (e.getCause() instanceof RuntimeException) throw e.getCause();
-      throw e;
+
+    private TracedSession(final Session session, Brave brave, SpanCollector collector) {
+        this.session = session;
+        this.brave = brave;
+        this.collector = collector;
+        this.clientTracer = brave.clientTracer();
+        this.useKeyspaceTimeout = session.getCluster().getConfiguration().getSocketOptions().getConnectTimeoutMillis();
+        this.version = session.getCluster().getConfiguration().getProtocolOptions().getProtocolVersion();
     }
-  }
-
-  @Override public void update(Host host, Statement statement, Exception e, long nanos) {
-    if (!(statement instanceof BoundStatement)) return;
-    Span span = cache.remove(statement);
-    if (span == null) {
-      if (statement.isTracing()) {
-        LOG.warn("{} not in the cache eventhough tracing is on", statement);
-      }
-      return;
+
+    @Override
+    public String getLoggedKeyspace() {
+        return session.getLoggedKeyspace();
     }
-    span.setDuration(nanos / 1000); // TODO: allow client tracer to end with duration
-    Endpoint local = span.getAnnotations().get(0).host; // TODO: expose in brave
-    long endTs = span.getTimestamp() + span.getDuration();
-    span.addToAnnotations(Annotation.create(endTs, "cr", local));
-    if (e != null) {
-      span.addToBinary_annotations(BinaryAnnotation.create(Constants.ERROR, e.getMessage(), local));
+
+    @Override
+    public Session init() {
+        return session.init();
     }
-    int ipv4 = ByteBuffer.wrap(host.getAddress().getAddress()).getInt();
-    Endpoint endpoint = Endpoint.create("cassandra3", ipv4, host.getSocketAddress().getPort());
-    span.addToBinary_annotations(BinaryAnnotation.address("sa", endpoint));
-    collector.collect(span);
-  }
-
-  @Override
-  public boolean equals(Object obj) {
-    if (obj instanceof TracedSession) {
-      TracedSession other = (TracedSession) obj;
-      return target.equals(other.target);
+
+    @Override
+    public ResultSetFuture executeAsync(final Statement statement) {
+        // don't trace any already self-traced spans
+        boolean alreadySelfTraced = null != statement.getOutgoingPayload()
+                && statement.getOutgoingPayload().containsKey(Cassandra3Storage.SELF_TRACING_KEY);
+
+        // o.a.c.tracing.Tracing.newSession must use the same format for the key zipkin
+        boolean supported = version.compareTo(ProtocolVersion.V4) >= 0;
+        if (supported && !alreadySelfTraced) {
+            useKeyspace(session, statement.getKeyspace());
+            SpanId spanId = null;
+            spanId = clientTracer.startNewSpan(statement.toString());
```

Comment:
please carry-forward the comments here and elsewhere unless they are invalid now. Ex. this one had..

```
// via an internal class z.s.cassandra3.NamedBoundStatement, toString() is a nice name
```

We don't want default BoundStatement toString as it breaks most advice about span naming